src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs

```diff
@@ -476,29 +476,61 @@ private void AddCentralTransitiveDependencyGroupsForPackageReference(PackageSpec
 
             // Do not pack anything from the runtime graphs
             // The runtime graphs are added in addition to the graphs without a runtime
-            foreach (var targetGraph in targetGraphs.Where(targetGraph => string.IsNullOrEmpty(targetGraph.RuntimeIdentifier)))
+            foreach (RestoreTargetGraph targetGraph in targetGraphs.Where(targetGraph => string.IsNullOrEmpty(targetGraph.RuntimeIdentifier)))
             {
-                var centralPackageVersionsForFramework = project.TargetFrameworks.Where(tfmi => tfmi.FrameworkName.Equals(targetGraph.Framework)).FirstOrDefault()?.CentralPackageVersions;
+                TargetFrameworkInformation targetFrameworkInformation = project.TargetFrameworks.FirstOrDefault(i => i.FrameworkName.Equals(targetGraph.Framework));
+
+                if (targetFrameworkInformation == null)
+                {
+                    continue;
+                }
 
                 // The transitive dependencies enforced by the central package version management file are written to the assets to be used by the pack task.
                 IEnumerable<LibraryDependency> centralEnforcedTransitiveDependencies = targetGraph
-                    .Flattened
-                    .Where(graphItem => graphItem.IsCentralTransitive && centralPackageVersionsForFramework?.ContainsKey(graphItem.Key.Name) == true)
-                    .Select((graphItem) =>
+                    .Graphs
+                    .SelectMany(i => i.InnerNodes)
+                    .Where(node => node.Item.IsCentralTransitive && targetFrameworkInformation.CentralPackageVersions.ContainsKey(node.Item.Key.Name))
+                    .Select((node) =>
                     {
-                        CentralPackageVersion matchingCentralVersion = centralPackageVersionsForFramework[graphItem.Key.Name];
+                        CentralPackageVersion matchingCentralVersion = targetFrameworkInformation.CentralPackageVersions[node.Item.Key.Name];
                         Dictionary<string, LibraryIncludeFlags> dependenciesIncludeFlags = _includeFlagGraphs[targetGraph];
 
+                        LibraryIncludeFlags suppressParent = LibraryIncludeFlags.None;
+
+                        if (project.RestoreMetadata.CentralPackageTransitivePinningEnabled)
+                        {
+                            // Centrally pinned dependencies are not directly declared but the PrivateAssets from the top-level dependency that pulled it in should apply to it also
+                            foreach (GraphNode<RemoteResolveResult> parentNode in FlattenParentNodes(node))
```

Comment:
This logic currently flattens all of the parents and goes through them top -> down.  I don't think `ParentNodes` ever contains more than one item from what I can tell.  This should result in the PrivateAssets from the actual top-level dependency just getting copied (unless we ever update the code to have multiple parents)